### PR TITLE
Refactor: evm-circuit util/chips to "gadgets" crate

### LINKED ISSUE #463: Migrate `util` module gadgets to the gadgets workspace member
Some generic gadgets like LTChip being worked on https://github.com/appliedzkp/zkevm-circuits/pull/460 are basically pulling other functions from `zkevm-circuits` such as `from_bytes`. 

This ends up with circular dependencies.

Therefore we should migrate as many non-zkevm-circuits specific gadgets into the new workspace member added in #457

cc: @roynalnaruto

### PR DESCRIPTION
**Pull request by roynalnaruto**
Closes #463 

- [x] Migrates most of `util` module (`Expr` trait and implementations for common types) to `gadgets` member crate
- [x] Fix compilation in `zkevm-circuits` member crate wherever appropriate
- [x] Adds `LessThanChip` to the `gadgets` member crate, to be used in #584
- [x] Migrate `binary_number` module from `zkevm_circuits::state_circuit` to `gadgets` crate
- [x] Migrate helper modules like `and`, `sum`, `or`, `not` and so on to `gadgets` crate

### CODE REVIEW

**Comment by roynalnaruto** [comment]
@CPerezz added this PR to address #463 . Will rebase the #584 PR once this is merged. Would be good to get a review on this.

**Comment by lispc** [comment]
i have another idea: reimport (pub use ...) gadgets::util::Expr as crate::util::Expr. So we don't have to modify lots of files, so the change will be much less, and some pending PR will be easier to merge conflicts.  what do you think? @roynalnaruto

**Comment by roynalnaruto** [comment]
@lispc great idea, yes it will make the changes minimal. I've made that change.

File: gadgets/src/less_than.rs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    ) -> Result<(), Error>;
+}
+
+/// Config for the Lt chip.
+#[derive(Clone, Debug)]
+pub struct LtConfig<F, const N_BYTES: usize> {
+    /// Denotes the lt outcome. If lhs < rhs then lt == 1, otherwise lt == 0.
+    pub lt: Column<Advice>,
+    /// Denotes the bytes representation of the difference between lhs and rhs.
+    pub diff: [Column<Advice>; N_BYTES],
```
Comment (han0110) [inline gadgets/src/less_than.rs:37]: Is that intended to not check the range of each byte in this config (user should do that somewhere else)? If so, it'd be nice to add some note here.

```suggestion
    /// Denotes the bytes representation in little-endian of the difference between lhs and rhs.
    /// Note that the range of each byte is not checked by this config.
    pub diff: [Column<Advice>; N_BYTES],
```

**Review by han0110 — APPROVED** [review]
LGTM. Nice refactor!

File: gadgets/src/less_than.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+impl<F: Field, const N_BYTES: usize> LtConfig<F, N_BYTES> {
+    /// Returns an expression that denotes whether lhs < rhs, or not.
+    pub fn is_lt(&self, meta: &mut VirtualCells<F>, rotation: Option<Rotation>) -> Expression<F> {
+        meta.query_advice(self.lt, rotation.unwrap_or_else(Rotation::cur))
+    }
+}
+
+/// Chip that compares lhs < rhs.
+#[derive(Clone, Debug)]
+pub struct LtChip<F, const N_BYTES: usize> {
```
Comment (ed255) [inline gadgets/src/less_than.rs:51]: Could you add a simple test for this Chip?

**Review by ed255 — APPROVED** [review]
LGTM! Thanks for working on this refactor :)

**Review by z2trillion — APPROVED** [review]
Looks good

**Pull request merged**